src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java

```diff
@@ -61,17 +64,17 @@ public static void initialize() {
         }
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
         if(!initialized)
-            initialize();
-        return authenticationProviders.get(scheme);
+            initialize(zks);
+        return WrappedAuthenticationProvider.wrap(authenticationProviders.get(scheme));
```

Comment:
OK - seems like a good set of abstractions to me. Let me know how you'd like it to change.